text/0000-partial_types.md

```diff
@@ -0,0 +1,395 @@
+
+Partial Types extension gives to Product Types (`PT = T1 and T2 and T3 and ..`), Structs and Tuples first of all, a good **mathematical guarantee** to borrow-checker that borrowing the whole variable with partial type and pretending to borrow just permitted fields is **fully safe** (without using `unsafe`).
+```rust
+struct StructABC { a: u32, b: i64, c: f32, }
+
+// function with partial parameter Struct
+fn ref_a (s : & StructABC.{a}) -> &u32 {
+    &s.a
+}
+
+let s = StructABC {a: 4, b: 7, c: 0.0};
+
+// partial expression at argument
+let sa = ref_a(& s.{a});
+```
+
+And since it is a guarantee by **type**, not by **values**, it has _zero cost_ in binary! Any type error is a compiler error, so no errors in the runtime.
+
+This extension is not only fully backward-compatible, but is fully forward-compatible! Forward-compatibility is an ability to use updated functions old way.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Partiality of type (or partial type access) is written as `Path.{fld1, fld2, fld3}` after Path (Type name), where `fld1`, `fld2`, .. are only permitted read (and maybe to write if variable is `mut`) fields of this type, the rest of fields are forbidden to read and write and unset.
+
+But the same time fields that are forbidden to read and write it is totally Ok to borrow, re-borrow, move, re-move, without any consequences - because the Compiler guarantee that in safe mode it is impossible to use such fields. It is a compile error if someone try to access it.
+
+## Partial Structs and Tuples
+
+For Product Types `PT = T1 and T2 and T3 and ..`), for structs, tuples we need not only partiality of a type, but also "partial access" expression: `Expr .{fld1, fld2, fld3}`, where `fld1`, `fld2`, .. are permitted fields of this type, the rest of fields are forbidden.
+
+One step to partial borrows Structs and Tuples.
+```rust
+struct Point {
+    x: f64,
+    y: f64, 
+    was_x: f64, 
+    was_y: f64,
+    state : f64,
+}
+let mut p1 = Point {x:1.0, y:2.0, was_x: 4.0, was_y: 5.0, state: 12.0};
+    // p1 : Point
+
+let ref_p1was = &mut p1.{wax_x, was_y};
+    // ref_p1was : &mut Point.{was_x, was_y}
+
+let ref_p1now = &mut p1.{x, y};
+    // ref_p1now : &mut Point.{x, y}
+```
+It is simple and will be possible. 
+
+It is easy to write functions, which consume partial parameters:
+```rust
+impl Point {
+    fn ref_x (self : & Self.{x}) -> &f64 {
+        &self.x
+    }
+
+    fn refmut_y (self : &mut Self.{y}) -> &mut f64 {
+        &mut self.y
+    }
+}
+let ref_p1x = p1.ref_x();
+let refmut_p1y = p1.refmut_y();
+```
+It is expected, that `self` is **always** cut partiality of argument by same partiality as self-parameter by partial expression before use (even if implicit rules are off)!
```

Comment:
the way you worded it is hard for me to understand...maybe this is better?
```suggestion
Here, since the methods' `self` types are partial references, only the needed fields are borrowed, so the call to `refmut_y` doesn't invalidate `ref_p1x`.
```
maybe also add something which uses the returned values so it's more clear they haven't been invalidated?

Reply:
i had meant that you should replace that sentence with what I proposed, since the original sentence is imo very confusing.